lib/client/src/jup.rs

```diff
@@ -0,0 +1,277 @@
+            max_quote_lots_including_fees,
+            &self.market,
+            self.oracle_price,
+            self.timestamp,
+        )?;
+        let (in_amount, out_amount) = match side {
+            Side::Bid => (
+                order_amounts.total_quote_taken_native,
+                order_amounts.total_base_taken_native,
+            ),
+            Side::Ask => (
+                order_amounts.total_base_taken_native,
+                order_amounts.total_quote_taken_native,
+            ),
+        };
+
+        Ok(Quote {
+            in_amount,
+            out_amount,
+            fee_mint: self.market.quote_mint,
+            fee_amount: order_amounts.fee,
+            ..Quote::default()
+        })
+    }
+
+    fn get_swap_and_account_metas(&self, swap_params: &SwapParams) -> Result<SwapAndAccountMetas> {
+        let SwapParams {
+            destination_mint,
+            source_mint,
+            user_destination_token_account,
+            user_source_token_account,
+            user_transfer_authority,
+            ..
+        } = swap_params;
+
+        let (side, base_account, quote_account) = if source_mint == &self.market.quote_mint {
+            (
+                JupiterSide::Bid,
+                user_destination_token_account,
+                user_source_token_account,
+            )
+        } else {
+            (
+                JupiterSide::Ask,
+                user_source_token_account,
+                user_destination_token_account,
+            )
+        };
+
+        let mut account_metas = vec![
+            AccountMeta::new(*user_transfer_authority, true),
+            AccountMeta::new(self.key, false),
+            AccountMeta::new(self.market.bids, false),
+            AccountMeta::new(self.market.asks, false),
+            AccountMeta::new(self.market.event_queue, false),
+            AccountMeta::new(self.market.base_vault, false),
+            AccountMeta::new(self.market.quote_vault, false),
+            AccountMeta::new(*base_account, false),
+            AccountMeta::new(*quote_account, false),
+            AccountMeta::new_readonly(Token::id(), false),
+            AccountMeta::new_readonly(System::id(), false),
+        ];
+        if self.market.oracle_a.is_some() && self.market.oracle_b.is_some() {
+            account_metas.extend(vec![
+                AccountMeta::new_readonly(self.market.oracle_a.key, false),
+                AccountMeta::new_readonly(self.market.oracle_b.key, false),
+            ]);
+        } else if self.market.oracle_a.is_some() {
+            account_metas.extend(vec![AccountMeta::new_readonly(
+                self.market.oracle_a.key,
+                false,
+            )]);
+        };
+
+        Ok(SwapAndAccountMetas {
+            swap: Swap::Openbook { side: { side } },
+            account_metas,
+        })
+    }
+
+    fn clone_amm(&self) -> Box<dyn Amm + Send + Sync> {
+        Box::new(self.clone())
+    }
+}
+
+pub struct AccountOracle<'a> {
```

Comment:
I don't think you need to reimplement this, `accounts_zerocopy` already exports a lot of helpers